docs: use_render_queue, use_liveness_scope, use_table_listener docs (#1044)
deephaven/deephaven-plugins: plugins/ui/docs/hooks/use_table_listener.md

# use_table_listener

> `use_table_listener` lets you listen to a table for updates. This is useful when you want to listen to a table and perform an action when the table updates.

Reviewer: The "this is an advanced feature" caveat should be mentioned in the paragraph, IMO. Stop someone in their tracks before getting to UI recommendations.

> `use_table_listener` lets you listen to a table for raw updates. This is an advanced feature requiring an understanding of how [table listeners](https://deephaven.io/core/docs/how-to-guides/table-listeners-python/) work and limitations of running code while the Update Graph is running. Most usages of this are more appropriate to implement with [the table data hooks](./overview.md#data-hooks). This is useful when you want to listen to the raw updates from a table and perform a custom action when the table updates.

Reviewer: ```suggestion
`use_table_listener` lets you listen to a table for raw updates. This is an advanced feature requiring an understanding of how [table listeners](https://deephaven.io/core/docs/how-to-guides/table-listeners-python/) work and the limitations of running code while the Update Graph is running. Most usages of this are more appropriate to implement with [table data hooks](./overview.md#data-hooks). This is useful when you want to listen to the raw updates from a table and perform a custom action when the table updates.
```

## Example

```python
from deephaven import time_table, ui
from deephaven.table import Table

_source = time_table("PT1s").update("X = i")

@ui.component
def ui_table_monitor(t: Table):
    def listener_function(update, is_replay):
        print(f"Table updated: {update}, is_replay: {is_replay}")

    ui.use_table_listener(t, listener_function, [])
    return t

table_monitor = ui_table_monitor(_source)
```

## UI recommendations

> 1. **Use table data hooks instead when possible**: `use_table_listener` is an advanced feature, requiring understanding of how the [table listeners](https://deephaven.io/core/docs/how-to-guides/table-listeners-python/) work, and limitations of running code while the Update Graph is running. Most usages of this are more appropriate to implement with [the table data hooks](./overview.md#data-hooks).

Reviewer: ```suggestion
1. **Use table data hooks instead when possible**: `use_table_listener` is an advanced feature requiring an understanding of how [table listeners](https://deephaven.io/core/docs/how-to-guides/table-listeners-python/) work and limitations of running code while the Update Graph is running. Most usages of this are more appropriate to implement with [the table data hooks](./overview.md#data-hooks).
```

> 1. **Use table data hooks instead when possible**: `use_table_listener` is an advanced feature requiring an understanding of how [table listeners](https://deephaven.io/core/docs/how-to-guides/table-listeners-python/) work and limitations of running code while the Update Graph is running. Most usages of this are more appropriate to implement with [the table data hooks](./overview.md#data-hooks).

Reviewer: ```suggestion
1. **Use table data hooks instead when possible**: `use_table_listener` is an advanced feature requiring an understanding of how [table listeners](https://deephaven.io/core/docs/how-to-guides/table-listeners-python/) work and limitations of running code while the Update Graph is running. Most usages of this are more appropriate to implement with [table data hooks](./overview.md#data-hooks).
```

Reviewer: Do we need this note repeated? In the intro seems fine?

## Display the last updated row

Here's an example that listens to table updates and will display the last update as a header above the table. This is a simple example to demonstrate how to use `use_table_listener` to listen to table updates and update state in your component.

```python
from deephaven import time_table, ui
from deephaven.table import Table

@ui.component
def ui_show_last_changed(t: Table):
    last_change, set_last_change = ui.use_state("No changes yet.")

    def listener_function(update, is_replay):
        set_last_change(f"{update.added()['X'][0]} was added")

    ui.use_table_listener(t, listener_function, [])
    return [ui.heading(f"Last change: {last_change}"), t]

_source = time_table("PT5s").update("X = i")
show_last_changed = ui_show_last_changed(_source)
```

## Display a toast

> Here's an example that listens to table updates and will display a toast message when the table updates. This is a simple example to demonstrate how to use `use_table_listener` to listen to table updates and display a toast message. Note you must use a [render queue](./use_render_queue.md) to trigger the toast, as the listener is not fired on the render thread.

Reviewer: ```suggestion
Here is a simple example that listens to table updates and displays a toast message when the table updates. Note you must use a [render queue](./use_render_queue.md) to trigger the toast, as the listener is not fired on the render thread.
```
